Announce (auto-refreshing) Official CVE Feed alpha (#35608)
kubernetes/website: content/en/blog/_posts/2022-08-26-Announcing-Auto-Refreshing-Official-CVE-Feed/index.md

---
layout: blog 

> title: "Announcing the Auto-refreshing Official Kubernetes CVE
> Feed"

Reviewer: ```suggestion
title: Announcing the Auto-refreshing Official Kubernetes CVE Feed
```

> date: 2022-08-26 

Reviewer: Please edit the path of this file and also change this field to match the assigned publication date: 2022-09-12

Author: done

slug: k8s-cve-feed-alpha
---

**Author**: Pushkar Joglekar (VMware)

> A programmatic way to keep track of Kubernetes CVEs for End Users, has been a
> longstanding request for the Kubernetes Community. With the release of v1.25,

Reviewer: ```suggestion
A long-standing request from the Kubernetes community has been to have
a programmatic way for end users to keep track of Kubernetes security issues
(CVEs, named after the database that tracks public security issues across different
products and vendors).
Accompanying the release of Kubernetes v1.25,
```

  Interest Groups and Committees.

## Implementation Details

An auto-refreshing CVE feed allows End Users to programmatically fetch the list
of CVEs and allows them to get the latest information from Kubernetes Community.

The feed is updated to ensure freshness and minimal maintainer overhead
by fetching the CVE related information through existing SRC workflow of
creating a GitHub Issue when announcing new Kubernetes CVEs.

### Pre-requisites

Until December 2021, it was not possible to filter for issues or PRs that are
tied to CVEs announced by Kubernetes SRC. To address that a new
label `official-cve-feed` was added. Relevant issues were then labelled with this
new label. The in-scope issues are `closed` issues for which
there is a CVE ID(s) and is officially announced as a Kubernetes security
vulnerability by SRC. You can now filter on all of these issues and
find them 
[here](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+is%3Aclosed+label%3Aofficial-cve-feed+).

For future security vulnerabilities, this label was added to the SRC playbook
such that all the future in-scope issues will automatically have this label.

> ### Prow, Cloud Buckets and Netlify - ftw

Reviewer: ```suggestion
### Prow, Cloud Buckets and Netlify {#implementation}
```

Author: > You know what, I'd split this into two articles:
> 
> * https://blog.k8s.io/ article: **Announcing the Official Kubernetes CVE Feed**
> * https://k8s.dev/blog/ article: **How I Implemented the Kubernetes Security Issues Feed**
> 
> What do you think @PushkarJ ? If we do that, both articles can go out on the same day and link to each other.

That's an interesting idea. Which sections in the current file would you recommend moving to the second blog on k8s.dev ?

Reviewer: > Which sections in the current file would you recommend moving to the second blog on k8s.dev ?

I'd shift all of the following to the contributor blog article:

```markdown
## Implementation Details

…

### Pre-requisites

…

### Prow, Cloud Buckets and Netlify 

…

## Design Considerations

…

### Integrity and Access Control Protections

…

#### GitHub Issues for Publicly announced CVEs

…

#### Prow configuration

…

#### Google Cloud Bucket

…

#### Website data templates
```

I would put the same
```markdown
### Freshness guarantees
```
heading in **both** articles, but with different content: outcome vs. implementation.

Author: @sftim makes sense to me. This would require some additional work though, so want to be cognizant of any deadlines. Are we okay if we post the feature blogs few days after v1.25 is GA (Aug 26) ? I don't expect to be able to work on this refactor until Monday Aug 22

Reviewer: I recommend that you focus on the announcement article - keep it simple - and work on the contributor article if and as time allows.
You already planned to have the announcement go out _after_ the new feed goes live, right?

approval process mentioned earlier.

#### Google Cloud Bucket

Write access to Google Cloud bucket is configured to be restricted to a set of
trusted community members managed via an
invite-only [Google Groups Membership](https://github.com/kubernetes/k8s.io/blob/main/groups/sig-security/groups.yaml)
under the `kubernetes.io` domain.

#### Website Data templates

Website data templates that fetch and parse the stored JSON blob, are managed under `kubernetes/website` and have to follow
the same code review and approval process as mentioned earlier.

### Freshness Guarantees

The job runs every `two` hours and compares the `sha256` checksum of the
existing contents of the bucket with checksum of the latest JSON blob generated through GitHub issues.
If the there is new data available the hashes do not match (typically because of
a newly announced CVE) and the updated JSON blob is pushed onto to the bucket
replacing the old blob and old hash checksum.

If the hashes match, the `write to bucket` operation is skipped to reduce
redundant updates to the cloud bucket. This also sets us up for more frequent runs

> of the prow job if needed in the future.

Reviewer: How would you explain this to someone who isn't particularly technical and just wants to know if the feed might be stale?

Author: Tried resolving this in the other PR https://github.com/kubernetes/contributor-site/pull/330

## What's Next?

> As we move towards graduation of this feature to `beta` from `alpha`, we are gathering feedback from end users who are using this feed.

Reviewer: ```suggestion
As we move towards graduation of this feature from alpha to beta, SIG Security are gathering feedback from end users who are using this alpha feed.
```
(emphasize that it's still alpha)

So in order to improve the feed in future Kubernetes Releases, if you have any feedback, please let us know by adding a comment to this [tracking issue](https://github.com/kubernetes/sig-security/issues/1) or
on #sig-security-tooling Kubernetes Slack channel.

If you would like to get involved in future iterations of this feed or would
love to work on other security relevant work please consider
joining [Kubernetes SIG Security](https://github.com/kubernetes/community/tree/master/sig-security#contact)
by joining our bi-weekly meetings or hanging out with us on our Slack Channel. 

> > A special shout out and massive thanks to Neha Lohia
> (@nehalohia27) and Tim Bannister (@sftim) for their stellar collaboration for
> many months from "ideation to implementation" of this feature.

Reviewer: ```suggestion
_A special shout out and massive thanks to Neha Lohia
(@nehalohia27) and Tim Bannister (@sftim) for their stellar collaboration for
many months from "ideation to implementation" of this feature._
```

Thanks. Please use italics as this isn't a quote.
